File: CSharp/Library/ArraySearch.cs
namespace Library
{
    public static class ArrayExtensions
    {
        public static int SearchIndex(
            this int[] ar,
            int item, int start, int end)
        {
            if (ar[start] > item) return start - 1;
            if (ar[end] < item) return end + 1;
            
            var mid = (end + start) / 2;
            return ar[mid] switch {
                var x when x == item => mid,
                var x when x < item  => ar.SearchIndex(item, mid + 1, end),
                _                    => ar.SearchIndex(item, start, mid - 1)
            };
        }
    }
}

File: FSharp/Library/ArraySearch.fs
namespace Library

module ArraySearch =
    let rec searchIndex (ar: int list) item startIndex endIndex =
        if ar.[startIndex] > item then
            startIndex - 1
        elif ar.[endIndex] < item then 
            endIndex + 1
        else
            let mid = (endIndex + startIndex) / 2
            match ar.[mid] with
            | x when x = item -> 
                mid
            | x when x < item -> 
                searchIndex ar item (mid + 1) endIndex
            | _ -> 
                searchIndex ar item startIndex (mid - 1)
